File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {TreeNode}
     */
    // BFS
    invertTree(root) {
        if(root == null){
            return null;
        }

        const queue = [root];
        while(queue.length){
            const node = queue.shift();
            [node.left,node.right] = [node.right,node.left];
            if(node.left) queue.push(node.left);
            if(node.right) queue.push(node.right);
        }
        return root;
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    bool sameTree(TreeNode* root , TreeNode* subRoot){
        if(!root and !subRoot){
            return true;
        }

        if(root and subRoot and root->val == subRoot->val){
            return sameTree(root->left, subRoot->left) and sameTree(root->right,subRoot->right); 
        }
        else return false;
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!subRoot) return true;
        if(!root) return false;

        if(sameTree(root,subRoot)) return true;
        
        return(isSubtree(root->left,subRoot) or isSubtree(root->right, subRoot));
    }
};
